### PR TITLE
Added in custom unitary to qasm

### LINKED ISSUE #3066: QasmError when parsing quantum volume qc.qasm()
### Information

- **Qiskit Terra version**: 0.9.0.dev0+de3f909
- **Python version**: 3.7.4
- **Operating system**: Windows Subsystem for Linux

### What is the current behavior?
If you pass a quantum volume circuit's qc.qasm() into qc.from_qasm_str, which calls qiskit's qasm parser, throws a QasmError.  The same error occurs for other circuits besides quantum volume.

### Steps to reproduce the problem
import qiskit.ignis.verification.quantum_volume as qv
qcs = qv.qv_circuits(qubit_lists=[[0,1]])
qc = qcs[0][0][0]
print(qc.qasm() + "\n")
new_qc = qc.from_qasm_str(qc.qasm())

### What is the expected behavior?
Not throwing an error and parsing correctly.

### PR DESCRIPTION
**Pull request by maddy-tod**
### Summary
Part of #3066 

Custom unitaries were not being written correctly to qasm. I have implemented a method that uses the `definition` of the unitary to create a corresponding new a gate in qasm. This is then added to the overall qasm output.

This does not fully fix #3066 as there is still the issue reported in #1566 which is that custom gates can't be read in from qasm. This change just ensures that we are generating valid qasm output.

### Details and comments
In `QuantumCircuit.qasm()` there is a small change to ensure that the gate definition of the unitary only gets written to the qasm once. This is done by the gate itself keeping track of whether it has written its definition yet, and then this variable being reset at the end of the call to `QuantumCircuit.qasm()`.

### CODE REVIEW

File: qiskit/extensions/unitary.py [hunk 1]
```diff
@@ -106,6 +111,55 @@ def _define(self):
...
+        """
+        # if this is true then we have written the gate definition already
+        # so we only need to write the name
+        if self.qasm_def_written:
+            return self._qasmif(self.qasm_name)
+
+        # we have worked out the definition before, but haven't written it yet
+        # so we need to write definition + name
+        if self.qasm_definiton:
+            self.qasm_def_written = True
```
Comment (kdk) [inline qiskit/extensions/unitary.py]: Can this be `self._qasm_def_written`?

Comment (maddy-tod) [reply qiskit/extensions/unitary.py]: I added 3 new fields, and they all should be private which I have done in 4d3aff9

File: qiskit/extensions/unitary.py [hunk 2]
```diff
@@ -62,6 +63,10 @@ def __init__(self, data, label=None):
         if input_dim != output_dim or 2**n_qubits != input_dim:
             raise ExtensionError(
                 "Input matrix is not an N-qubit operator.")
+
+        self._qasm_name = None
+        self._qasm_definiton = None
+        self._qasm_def_written = False
```
Comment (1ucian0) [inline qiskit/extensions/unitary.py]: Maybe `self._qasm_definiton` can be merged with `self._qasm_def_written`?

Comment (maddy-tod) [reply qiskit/extensions/unitary.py]: I thought about this but I decided to keep them separate. `self._qasm_definition` holds the definition of this gate in QASM, and `self._qasm_def_written` keep track of whether in this call to the method we have already written the QASM. By keeping them separate we don't have to regenerate the QASM definition of the gate with every call to `circuit.qasm()` and I think this should therefore offer a slight speedup.

Comment (1ucian0) [reply qiskit/extensions/unitary.py]: I think we do similar things in other parts with of the code with `@property`. Let me write come up with an example ...

Comment (1ucian0) [reply qiskit/extensions/unitary.py]: Also, why this caching is needed here and not in the rest of the gates?

Comment (maddy-tod) [reply qiskit/extensions/unitary.py]: Other gates can be written just using their name, so no caching is needed as nothing is being generated. This gate needs to be written as a new gate definition, so we have to generate the QASM definition so we use caching so we don't have to keep regenerating it every time.

File: test/python/circuit/test_unitary.py [hunk 3]
```diff
@@ -183,3 +183,26 @@ def test_labeled_unitary(self):
...
+        expected_qasm = "OPENQASM 2.0;\n" \
+                        "include \"qelib1.inc\";\n" \
+                        "qreg q0[2];\ncreg c0[1];\n" \
+                        "x q0[0];\n" \
+                        "gate custom_gate p0 {\n" \
+                        "\tu3(0.0,0.0,0.0) p0;\n" \
+                        "}\n" \
+                        "custom_gate q0[0];\n" \
+                        "custom_gate q0[1];\n"
+        self.assertEqual(expected_qasm, qc.qasm())
```
Comment (kdk) [inline test/python/circuit/test_unitary.py]: Can you add a test for writing the qasm of a circuit twice? And one for decomposing a two qubit unitary?

File: qiskit/circuit/quantumcircuit.py [hunk 4]
```diff
@@ -556,6 +557,12 @@ def qasm(self):
                 string_temp += "%s %s;\n" % (instruction.qasm(),
                                              ",".join(["%s[%d]" % (j.register.name, j.index)
                                                        for j in qargs + cargs]))
+            if instruction.name == 'unitary':
+                unitary_gates.append(instruction)
+
+        # this resets them, so if another call to qasm() is made the gate def is added again
+        for gate in unitary_gates:
+            gate.qasm_def_written = False
```
[the remark this answers is not in the document]
Comment (maddy-tod) [reply qiskit/circuit/quantumcircuit.py]: Done

File: test/python/circuit/test_unitary.py (hunk 3 above)
Comment (maddy-tod) [reply test/python/circuit/test_unitary.py]: I have added a test for writing the qasm of a circuit twice. However the decomposition for a 2 qubit unitary seems to be somewhat random (I believe due to [this algorithm](https://github.com/Qiskit/qiskit-terra/blob/998c232b00c56040a14af23999a07787c0090e5e/qiskit/quantum_info/synthesis/two_qubit_decompose.py#L164)) and so the qasm generated each time is ever so slightly different but equivalent. This means we cannot do a straightforward equivalence check of the new qasm. We also cannot test that the original circuit and the circuit from the loaded qasm generate the same results as you cannot load custom gates into the DAG (`qiskit.exceptions.QiskitError: 'Custom non-opaque gates are not supported by as_to_dag module'`), so I'm not sure there is a good way to test this part.

Comment (kdk) [reply test/python/circuit/test_unitary.py]: You could try a two qubit identity (`np.eye(4)`), which should reliably decompose to two `u3(0.0,0.0,0.0)`s. The issue on loading custom gates from qasm is tracked in #1566 .

Comment (maddy-tod) [reply test/python/circuit/test_unitary.py]: Thanks for the suggestion! Added in e0da9c7

**Pull request merged**